ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterSortPredicates.java

```diff
@@ -0,0 +1,268 @@
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexDynamicParam;
+import org.apache.calcite.rex.RexFieldAccess;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.rex.RexVisitorImpl;
+import org.apache.calcite.util.Pair;
+import org.apache.hadoop.hive.ql.optimizer.calcite.stats.FilterSelectivityEstimator;
+import org.apache.hadoop.hive.ql.optimizer.calcite.stats.HiveRelMdSize;
+
+
+/**
+ * Rule that sorts conditions in a filter predicate to accelerate query processing
+ * based on selectivity and compute cost. Currently it is not applied recursively,
+ * i.e., it is only applied to top predicates in the condition.
+ */
+public class HiveFilterSortPredicates extends RelOptRule {
+
+  public static final HiveFilterSortPredicates INSTANCE = new HiveFilterSortPredicates();
+
+
+  private HiveFilterSortPredicates() {
+    super(
+        operand(Filter.class,
+            operand(RelNode.class, any())));
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    final Filter filter = call.rel(0);
+
+    HiveRulesRegistry registry = call.getPlanner().getContext().unwrap(HiveRulesRegistry.class);
+
+    // If this operator has been visited already by the rule,
+    // we do not need to apply the optimization
+    if (registry != null && registry.getVisited(this).contains(filter)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    final Filter filter = call.rel(0);
+    final RelNode input = call.rel(1);
+
+    // Register that we have visited this operator in this rule
+    HiveRulesRegistry registry = call.getPlanner().getContext().unwrap(HiveRulesRegistry.class);
+    if (registry != null) {
+      registry.registerVisited(this, filter);
+    }
+
+    final RexNode originalCond = filter.getCondition();
+    RexSortPredicatesShuttle sortPredicatesShuttle = new RexSortPredicatesShuttle(
+        input, filter.getCluster().getMetadataQuery());
+    final RexNode newCond = originalCond.accept(sortPredicatesShuttle);
+    if (!sortPredicatesShuttle.modified) {
+      // We are done, bail out
+      return;
+    }
+
+    // We register the new filter so we do not fire the rule on it again
+    final Filter newFilter = filter.copy(filter.getTraitSet(), input, newCond);
+    if (registry != null) {
+      registry.registerVisited(this, newFilter);
+    }
+
+    call.transformTo(newFilter);
+  }
+
+  /**
+   * If the expression is an AND/OR, it will sort predicates accordingly
+   * to maximize performance.
+   * In particular, for AND clause:
+   * rank = (selectivity - 1) / cost per tuple
+   * Similarly, for OR clause:
```

Comment:
Can you add a comment to explain why OR is sorted differently?

Reply:
Added comments to `rankingAnd` and `rankingOr` methods.